### ISSUE TITLE
New MS lesion version requires N-fold averaging and custom trainer support

### ISSUE THREAD

**Issue by plbenveniste** [spinalcordtoolbox/spinalcordtoolbox#4995]
Hi!

I want to add the new MS lesion model.
However, after discussion with @jcohenadad, we have decided that it would be better to add the 5-fold model to SCT to:
- get better segmentation (compared to a 1-fold model)
- allow creation of soft segmentation from the average output of the 5 folds. 

My questions are the following:
- How much refactoring will this cause?
- This model having a custom trainer, what has been decided for now regarding this?

Thanks 😃

**Comment by joshuacwnewton** [comment]
I had implemented multi-fold averaging for the `ivadomed` models some time ago, so some of the infrastructure should still exist? I would have to dig into the code to check how difficult it would be to apply this to non-ivadomed models.

**Comment by plbenveniste** [comment]
Action task from SCT meeting: Look into whether or not nnunetv2 supports soft or if we can get the 5 predictions before binarization. 

It seems that there is no easy way to stop the model from merging the 5 folds predictions.

In SCT we use: `predictor.predict_single_npy_array`
Code structure: 
- folds are gathered [here](https://github.com/MIC-DKFZ/nnUNet/blob/8c4184d46b60059ff7dc8f74cd535e13554bdeca/nnunetv2/inference/predict_from_raw_data.py#L73-L74): in `initialize_from_trained_model_folder` with `use_folds = nnUNetPredictor.auto_detect_available_folds`
- model is initialized [here](https://github.com/MIC-DKFZ/nnUNet/blob/8c4184d46b60059ff7dc8f74cd535e13554bdeca/nnunetv2/inference/predict_from_raw_data.py#L67): with `initialize_from_trained_model_folder(`.
- prediction is done [here](https://github.com/MIC-DKFZ/nnUNet/blob/8c4184d46b60059ff7dc8f74cd535e13554bdeca/nnunetv2/inference/predict_from_raw_data.py#L450C9-L450C25): with `predicted_logits = self.predict_logits_from_preprocessed_data(dct['data']).cpu()`
- predictions are averaged if there are multiple folds ([here](https://github.com/MIC-DKFZ/nnUNet/blob/8c4184d46b60059ff7dc8f74cd535e13554bdeca/nnunetv2/inference/predict_from_raw_data.py#L499-L500)) in `predict_logits_from_preprocessed_data(`:
```python
 if len(self.list_of_parameters) > 1:
            prediction /= len(self.list_of_parameters) 
```

**Comment by joshuacwnewton** [comment]
Copying over a comment made by @naga-karthik for posterity:

> [Regarding the SCT meeting on 28/8] — I saw that there was a discussion on getting soft masks from nnunet inference. I tried something like this, maybe this could be a starting point https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/main/nnUnet/save_softmax_npz_as_nifti.py

**Comment by naga-karthik** [comment]
Thanks @joshuacwnewton for posting it here! There is IMPORTANT note though: my script does _not_ do N-fold averaging -- it's just taking the raw softmax probabilities and saving it as nifti file. Getting the raw probabilities does not have anything to do having N folds -- they can be obtained from a single fold also. There are two ways to get soft masks now:
1. Get the standard binary segmentations from a single fold (the usual inference from nnunet models in SCT) --> Repeat N times for N folds --> average to get soft masks 
2. Get the raw softmax probabilties (what my script does) --> Repeat N times --> average to get soft masks 
 
**Note:** (2) is slightly more accurate than (1). This is because in (1), we're already losing some softness by binarizing the prediction (what nnunet does automatically), and by averaging 5 folds, we're forcefully getting soft masks. In (2) however, we still keep the raw probabilities and averaging them N times, gives a slightly more accurate representation of the softness.

**Comment by plbenveniste** [comment]
I have opened an issue in the nnunet repo to see if they are planning on creating this feature: https://github.com/MIC-DKFZ/nnUNet/issues/2902

**Comment by plbenveniste** [comment]
## Investigation into the softness of the lesion segmentations
I am investigating how to improve the softness of the soft MS lesion segmentations. In this experiment, I am trying two strategies:
- 1st strategy: using the same method as nnUNet for converting the softmax probabilities to binary segmentation. They keep voxels in the binary segmentation if the probability of the voxel of class 1 (=lesion) is greater than the probability of class 0 (=background)
- 2nd strategy: using thresholding of the softmax output probabilities of class 1.

### Example on image `ms-multi-spine-challenge-2024/sub-045_T2w.nii.gz` on upper lesion

No predicted lesion for the binary segmentation: 

<img width="1485" height="738" alt="Image" src="https://github.com/user-attachments/assets/fb1833c6-e083-47ba-a05d-dcc1e43de626" />

### Example on image `ms-multi-spine-challenge-2024/sub-045_T2w.nii.gz` on lower lesion

With predicted lesion in the binary segmentation:

<img width="1500" height="728" alt="Image" src="https://github.com/user-attachments/assets/e0bbff99-93cc-4c6f-99df-0ba89fbe83b4" />

### Conclusion

I would not use the strategy used by nnunet. Instead, I would use the thresholding method. It brings more softness to the soft-segmentation. At 1e-8, we can see some pixels appearing in weird areads of the image. Therefore, I would recommend selecting smaller threshold between 1e-4 and 1e-6.

**Comment by naga-karthik** [comment]
>  At 1e-8, we can see some pixels appearing in weird areads of the image. Therefore, I would recommend selecting smaller threshold between 1e-3 and 1e-5.

These are super small values and with such low thresholds it feels like we're forcing softness from the model. What about higher values in the range of 0.1-0.5?

**Comment by plbenveniste** [comment]
> These are super small values and with such low thresholds it feels like we're forcing softness from the model. What about higher values in the range of 0.1-0.5?

I agree that these are very small values. However, is it a problem to "force the softness"? Also, if we use values between 0.1-0.5 the soft segmentation will basically be the same as the binary segmentation... 

Another strategy is being explored. Give me a few minutes to add it...

**Comment by jcohenadad** [comment]
To me strategy 3 looks a lot like strategy 1, and the latter has the merit of not imposing an arbitrary threshold of 0.1 and 1e-5

**Comment by plbenveniste** [comment]
> To me strategy 3 looks a lot like strategy 1, and the latter has the merit of not imposing an arbitrary threshold of 0.1 and 1e-5

I agree with you. However, it has the downside of not being able to provide soft segmentation if the user decides to use only 1 fold. I suggest to do strategy A (this [comment](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4995#issuecomment-3339139889)) for each fold. This way, there is no use of thresholds and the user can still get soft segmentation when running on a single fold (eventhough the soft segmentation will likely be very close to the binary segmentation). I will implement and test this next week.
